File: sdk/typescript/src/releases/verify.ts
import type { VerifyReleaseDefaultsResponse } from "../generated/kms.js";

/** Bounded verdicts returned by verifyReleaseDefaults for one alias. */
export const VERIFY_VERDICTS = [
  "match",
  "differs",
  "missing_in_release",
  "unknown_alias",
  "secret_alias",
  "unsupported_content_type",
] as const;
export type VerifyVerdict = (typeof VERIFY_VERDICTS)[number];

/**
 * One parameter alias with the lowercase hex SHA-256 of its canonical encoded
 * value (`parameterHash`). Secret aliases must not be sent; the server reports
 * them as `secret_alias` without reading them.
 */
export interface VerifyDefaultsEntry {
  readonly alias: string;
  readonly contentType: string;
  readonly sha256: string;
}

/** One value-free comparison of source-owned defaults against an active release. */
export interface VerifyReleaseDefaultsOptions {
  /** The "env/app" whose active release is compared. Required: verification identities are typically unbound. */
  readonly namespace: string;
  /** Release name; empty selects the application's configured release name. */
  readonly release?: string;
  /** Informational label carried with the request. */
  readonly profile?: string;
  /** Generated contract schema digest; empty skips the schema check and leaves schemaMatches false. */
  readonly schemaSha256?: string;
  readonly entries: readonly VerifyDefaultsEntry[];
  readonly signal?: AbortSignal;
  readonly deadline?: Date;
}

/** The server's bounded verdict for one alias. */
export interface VerifyDefaultsVerdict {
  readonly alias: string;
  readonly verdict: VerifyVerdict;
}

/** Validated, value-free server response. */
export interface VerifyReleaseDefaultsResult {
  readonly releaseName: string;
  readonly releaseVersion: bigint;
  readonly activationRevision: bigint;
  readonly schemaMatches: boolean;
  readonly entries: readonly VerifyDefaultsVerdict[];
  readonly matchCount: number;
  readonly differsCount: number;
  readonly missingCount: number;
  readonly unknownAliasCount: number;
  readonly secretAliasCount: number;
  readonly unsupportedCount: number;
  /** Parameter aliases the release pins that the request did not mention. */
  readonly unverifiedCount: number;
  /** True when the schema matched and every entry matched. */
  passed(): boolean;
}

export function isVerifyVerdict(value: string): value is VerifyVerdict {
  return (VERIFY_VERDICTS as readonly string[]).includes(value);
}

export function validLowerHex64(value: string): boolean {
  return /^[0-9a-f]{64}$/u.test(value);
}

/** @internal Validate the wire response against the request and freeze the result. */
export function verifyResultFromWire(
  response: VerifyReleaseDefaultsResponse,
  requested: ReadonlySet<string>,
  fail: (message: string) => never,
): VerifyReleaseDefaultsResult {
  if (response.entries.length !== requested.size) {
    fail(`verify response has ${response.entries.length} verdicts for ${requested.size} entries`);
  }
  const tally = new Map<VerifyVerdict, number>();
  const seen = new Set<string>();
  const entries: VerifyDefaultsVerdict[] = [];
  for (const [index, entry] of response.entries.entries()) {
    if (!entry) fail(`verify response entry ${index} is empty`);
    if (!requested.has(entry.alias) || seen.has(entry.alias)) {
      fail("verify response names an unknown or duplicated alias");
    }
    if (!isVerifyVerdict(entry.verdict)) fail("verify response entry has an invalid verdict");
    seen.add(entry.alias);
    tally.set(entry.verdict, (tally.get(entry.verdict) ?? 0) + 1);
    entries.push(Object.freeze({ alias: entry.alias, verdict: entry.verdict }));
  }
  const counts: readonly (readonly [VerifyVerdict, number])[] = [
    ["match", response.matchCount],
    ["differs", response.differsCount],
    ["missing_in_release", response.missingInReleaseCount],
    ["unknown_alias", response.unknownAliasCount],
    ["secret_alias", response.secretAliasCount],
    ["unsupported_content_type", response.unsupportedContentTypeCount],
  ];
  for (const [verdict, count] of counts) {
    if (!validCount(count) || count !== (tally.get(verdict) ?? 0)) {
      fail(`verify response ${verdict} count disagrees with its verdicts`);
    }
  }
  if (!validCount(response.unverifiedCount)) fail("verify response unverified count is invalid");
  const schemaMatches = response.schemaMatches === true;
  const frozenEntries = Object.freeze(entries);
  return Object.freeze({
    releaseName: response.name,
    releaseVersion: response.version,
    activationRevision: response.activationRevision,
    schemaMatches,
    entries: frozenEntries,
    matchCount: response.matchCount,
    differsCount: response.differsCount,
    missingCount: response.missingInReleaseCount,
    unknownAliasCount: response.unknownAliasCount,
    secretAliasCount: response.secretAliasCount,
    unsupportedCount: response.unsupportedContentTypeCount,
    unverifiedCount: response.unverifiedCount,
    passed: () => schemaMatches && frozenEntries.every((entry) => entry.verdict === "match"),
  });
}

function validCount(value: number): boolean {
  return Number.isInteger(value) && value >= 0;
}

File: sdk/python/kms_paramstore/configstore/verify.py
"""Hash-only defaults verification against an active KMS release."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from .canonical import parameter_hash
from .contract import ContractEntry

_VERDICTS = (
    "match", "differs", "missing_in_release", "unknown_alias",
    "secret_alias", "unsupported_content_type",
)


@dataclass(frozen=True)
class VerifyEntryResult:
    alias: str
    content_type: str
    verdict: str


@dataclass(frozen=True)
class VerifyResult:
    namespace: str
    release_name: str
    release_version: int
    activation_revision: int
    schema_matches: bool
    entries: tuple[VerifyEntryResult, ...]
    unverified: int = 0

    @property
    def passed(self) -> bool:
        return self.schema_matches and all(entry.verdict == "match" for entry in self.entries)

    def failures(self) -> tuple[VerifyEntryResult, ...]:
        return tuple(sorted((entry for entry in self.entries if entry.verdict != "match"), key=lambda item: item.alias.encode()))

    def report(self) -> str:
        rows = [("VERDICT", "ALIAS", "CONTENT_TYPE")] + [
            (entry.verdict, entry.alias, entry.content_type)
            for entry in sorted(self.entries, key=lambda item: item.alias.encode())
        ]
        widths = [max(len(row[index]) for row in rows) + 2 for index in range(2)]
        lines = [f"{self.namespace} {self.release_name}@{self.release_version}#{self.activation_revision}  schema: {'match' if self.schema_matches else 'differs'}"]
        lines.extend(f"{row[0]:<{widths[0]}}{row[1]:<{widths[1]}}{row[2]}" for row in rows)
        counts = {verdict: sum(entry.verdict == verdict for entry in self.entries) for verdict in _VERDICTS}
        lines.append("summary: " + " ".join(f"{name}={counts[name]}" for name in _VERDICTS) + f" unverified={self.unverified}")
        lines.append("result: active release " + ("matches" if self.passed else "differs from") + " source defaults")
        return "\n".join(lines) + "\n"


def verify_defaults(
    client: object, *, namespace: str, schema_sha256: str,
    contract: tuple[ContractEntry, ...], groups: Mapping[str, str],
    release: str = "", profile: str = "", **options: Any,
) -> VerifyResult:
    if not namespace.strip():
        raise TypeError("configstore: verify requires namespace")
    entries, content_types = [], {}
    for item in contract:
        if item.kind != "parameter":
            continue
        if item.alias not in groups:
            raise ValueError(f"configstore: verify: missing encoded parameter group {item.alias}")
        entries.append({"alias": item.alias, "content_type": item.content_type, "sha256": parameter_hash(item.content_type, groups[item.alias])})
        content_types[item.alias] = item.content_type
    method = getattr(client, "verify_release_defaults")
    response = method(namespace=namespace, release=release, profile=profile, schema_sha256=schema_sha256, entries=entries, **options)
    results = tuple(VerifyEntryResult(item.alias, content_types.get(item.alias, ""), item.verdict) for item in response.entries)
    return VerifyResult(namespace, response.release_name, response.release_version, response.activation_revision, response.schema_matches, results, response.unverified_count)


async def verify_defaults_async(client: object, **kwargs: Any) -> VerifyResult:
    namespace, schema_sha256 = kwargs.pop("namespace"), kwargs.pop("schema_sha256")
    contract, groups = kwargs.pop("contract"), kwargs.pop("groups")
    if not isinstance(namespace, str) or not namespace.strip():
        raise TypeError("configstore: verify requires namespace")
    entries, content_types = [], {}
    for item in contract:
        if item.kind == "parameter":
            if item.alias not in groups:
                raise ValueError(
                    f"configstore: verify: missing encoded parameter group {item.alias}"
                )
            entries.append({"alias": item.alias, "content_type": item.content_type, "sha256": parameter_hash(item.content_type, groups[item.alias])})
            content_types[item.alias] = item.content_type
    response = await getattr(client, "verify_release_defaults")(namespace=namespace, schema_sha256=schema_sha256, entries=entries, **kwargs)
    results = tuple(VerifyEntryResult(item.alias, content_types.get(item.alias, ""), item.verdict) for item in response.entries)
    return VerifyResult(namespace, response.release_name, response.release_version, response.activation_revision, response.schema_matches, results, response.unverified_count)
